src/vsr/replica.zig

```diff
@@ -5254,26 +5290,59 @@ pub fn ReplicaType(
 
             self.set_op_and_commit_max(
                 do_view_change_head.op,
-                // `set_op_and_commit_max()` expects the highest commit_max that we know of.
-                // But DVCs include replica's `commit_min`, not `commit_max`.
-                std.math.max(
-                    self.commit_max,
-                    do_view_change_head.commit_min_max,
-                ),
+                commit_max: {
+                    var commit_max = self.commit_max;
+                    // `set_op_and_commit_max()` expects the highest commit_max that we know of.
+                    // But DVCs include replica's `commit_min`, not `commit_max`.
+                    commit_max = std.math.max(commit_max, do_view_change_head.commit_min_max);
+                    // An op cannot be uncommitted if it is definitely outside the pipeline.
+                    // Use `do_view_change_head.op` instead of `replica.op` since the former is
+                    // about to become the new `replica.op`.
+                    commit_max = std.math.max(
+                        commit_max,
+                        do_view_change_head.op -| constants.pipeline_max,
+                    );
+                    break :commit_max commit_max;
+                },
```

Comment:
I think this could actually be cleaner using two nested inline `max()` calls, plus comma to layout. This avoids introducing the local variable, and avoids the need for a block.